include/zephyr/posix/posix_signal.h

```diff
@@ -0,0 +1,398 @@
+#endif
+
+/* sig_atomic_t must be defined by the libc signal.h */
+
+#define SIGRTMIN 32
+#if defined(_POSIX_REALTIME_SIGNALS) || defined(__DOXYGEN__)
+BUILD_ASSERT(CONFIG_POSIX_RTSIG_MAX >= 0);
+#define SIGRTMAX (SIGRTMIN + CONFIG_POSIX_RTSIG_MAX)
+#else
+#define SIGRTMAX SIGRTMIN
+#endif
+
+#if !defined(_SIGSET_T_DECLARED) && !defined(__sigset_t_defined)
+typedef struct {
+	unsigned long sig[DIV_ROUND_UP(SIGRTMAX + 1, BITS_PER_LONG)];
+} sigset_t;
+#define _SIGSET_T_DECLARED
+#define __sigset_t_defined
+#endif
+
+#if !defined(_PID_T_DECLARED) && !defined(__pid_t_defined)
+typedef long pid_t;
+#define _PID_T_DECLARED
+#define __pid_t_defined
+#endif
+
+#if defined(_POSIX_THREADS) || defined(__DOXYGEN__)
+
+#if !defined(_PTHREAD_ATTR_T_DECLARED) && !defined(__pthread_attr_t_defined)
+typedef struct {
+	void *stack;
+	unsigned int details[2];
+} pthread_attr_t;
+#define _PTHREAD_ATTR_T_DECLARED
+#define __pthread_attr_t_defined
+#endif
+
+#endif
+
+#if defined(_POSIX_REALTIME_SIGNALS) || defined(__DOXYGEN__)
+
+union sigval; /* forward declaration (to preserve spec order) */
+
+#if !defined(_SIGEVENT_DECLARED) && !defined(__sigevent_defined)
+typedef struct {
+#if defined(_POSIX_THREADS) || defined(__DOXYGEN__)
+	pthread_attr_t *sigev_thread_attr;
+#endif
+	union sigval sigev_value;
+	int sigev_notify;
+	int sigev_signo;
+} sigevent_t;
+#define _SIGEVENT_DECLARED
+#define __sigevent_defined
+#endif
+
+#define SIGEV_NONE   1
+#define SIGEV_SIGNAL 2
+#define SIGEV_THREAD 3
+
+/* Signal constants are defined below */
+
+#endif /* defined(_POSIX_REALTIME_SIGNALS) || defined(__DOXYGEN__) */
+
+#if !defined(_SIGVAL_DECLARED) && !defined(__sigval_defined)
+union sigval {
+	int sival_int;
+	void *sival_ptr;
+};
+#define _SIGVAL_DECLARED
+#define __sigval_defined
+#endif
+
+/* SIGRTMIN and SIGRTMAX defined above */
+
+#if defined(_POSIX_REALTIME_SIGNALS) || defined(__DOXYGEN__)
+
+#if !defined(_SIGACTION_DECLARED) && !defined(__sigaction_defined)
+struct sigaction {
+	union {
+		void (*sa_handler)(int sig);
+		void (*sa_sigaction)(int sig, siginfo_t *info, void *context);
+	};
+	sigset_t sa_mask;
+	int sa_flags;
+};
+#define _SIGACTION_DECLARED
+#define __sigaction_defined
+#endif
+
+#define SIG_BLOCK   1
+#define SIG_UNBLOCK 2
+#define SIG_SETMASK 0
+
+#if defined(_XOPEN_SOURCE) || defined(__DOXYGEN__)
+#define SA_NOCLDSTOP 0x00000001
+#define SA_ONSTACK   0x00000002
+#endif
+#define SA_RESETHAND 0x00000004
+#define SA_RESTART   0x00000008
+#define SA_SIGINFO   0x00000010
+#if defined(_XOPEN_SOURCE) || defined(__DOXYGEN__)
+#define SA_NOCLDWAIT 0x00000020
+#endif
+#define SA_NODEFER  0x00000040
+#define SS_ONSTACK  0x00000001
+#define SS_DISABLE  0x00000002
+#define MINSIGSTKSZ 4096
+#define SIGSTKSZ    4096
+
+#if !defined(_MCONTEXT_T_DECLARED) && !defined(__mcontext_t_defined)
+typedef struct {
+	/* FIXME: there should be a much better Zephyr-specific structure that can be used here */
+	unsigned long gregs[32];
+	unsigned long flags;
+} mcontext_t;
```

Comment:
@andyross @npitre @ycsin - is there a better zephyr-specific struct that can be typedef'ed to `mcontext_t`?

We don't actually use this anywhere yet, and have no plans of supporting userspace context switching (which is obsolescent anyway), but it might be nice to be semi-accurate at the start.

Reply:
You'd need struct _callee_saved, struct _thread_arch and some of struct arch_esf. 
So if the intention is not to support swapcontext() then I'd simply leave it empty.